Episode V: The Locks Strike Back

jeremyk-91 opened the pull request:
Note: This is in a reasonable state for review, but I would like to manual-ETE test this on a reasonably complete stack of internal products before merging.

Is failing ETE tests because of #1809. We need to fix that first before merging this in (though I think this doesn't block reviewing this PR).

Also, I plan to break out the SerializableErrorDecoder to a separate PR so that we can address #1813 by itself. This PR will switch AtlasDbErrorDecoder over to use it.

**TODO BEFORE MERGING**

- [x] Merge #1831
- [x] Rebase the PR and get it into a state where it passes CI
- [x] Make the `ERROR_MARGIN` constant configurable
- [x] ETE Smoke Test on an internal stack
- [x] Verify using Jeremy's blocking CLI that the number of server threads remains sane

---

**Goals (and why)**:

* Solve #1680 without failing long-running lock requests, which blocks our big guy #1728 

**Implementation Description (bullets)**:

This PR has three nontrivial pieces. I would have preferred to have done them separately, though I want to verify they work in concert before merging them.

* `BlockingTimeLimitedLockService` actually time-limits the sensitive operations on the lock service.
* `BlockingTimeouts` parses the Dropwizard configuration and extracts the minimum blocking timeout across all application connectors. It then calculates the blocking timeout, which is currently 97 percent of the idle timeout. I've documented why it's safe in terms of correctness even if we have a race between this and the idle timeout.
* Changes to the exception handling logic - an exception is now returned with a body. If an old client receives a message this is fine, as it would previously have ignored the body. I made the old clients treat `BlockingTimeoutException` as non-fast-failover (no `Retry-After` header) to allow old clients to behave reasonably well with this in place.

This PR does NOT include the following (but I want to do these):

* ADR for how the lock retrying works
* Automated ETE test that verifies that a lock held for more than an idle timeout doesn't cause other clients to overload the server.

**Concerns (what feedback would you like?)**:

* I have only time-limited the LockService methods that throw `InterruptedException` as I guess they are the ones likely to be contentious. Is this reasonable and/or safe?
* `BlockingTimeouts` is rather unpleasant. Is there an easier way to enforce the key invariant (blocking timeout is a smidge under the minimum idle timeout) and/or should I even bother to enforce this?

**Where should we start reviewing?**:

* `BlockingTimeLimitedLockService` is probably a good starting point.

**Priority (whenever / two weeks / yesterday)**:

Not too urgent. I'm rather busy with non-AtlasDB related work this week, so my bandwidth is somewhat limited (though nonzero).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/1808)
<!-- Reviewable:end -->

hsaraogi commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/NotCurrentLeaderExceptionMapper.java:
```diff
@@ -29,6 +28,6 @@
     @Override
     public Response toResponse(NotCurrentLeaderException exception) {
-        return Response.noContent().status(503).header(HttpHeaders.RETRY_AFTER, "0").build();
+        return ExceptionMappers.encode503ResponseWithRetryAfter(exception);
```
Isnt this still `noContent` - not sure if entity is content but I checked and is not clear from the API?

jeremyk-91 replied:
Yep, the entity becomes the content of a message. This is added in `encode503ResponseInternal` (`ExceptionMappers:39`). I'll add some docs to clarify this

hsaraogi commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -63,6 +63,17 @@ develop
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1782>`__)
 
+    *    - |improved| |devbreak|
+         - The format of exception messages has been brought in line with that of the `palantir/http-remoting <https://github.com/palantir/http-remoting>`__ library.
```
May be more clear to mention which exception messges.

jeremyk-91 replied:
Sure, that's right. I should have specified that this has to do with remote exceptions.

hsaraogi commented on timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/lock/BlockingTimeoutsIntegrationTest.java:
```diff
@@ -0,0 +1,71 @@
+    public void returnsBlockingTimeoutWithConnectorWithIdleTimeouts() throws IOException {
+        assertThat(BlockingTimeouts.getBlockingTimeout(OBJECT_MAPPER, getConfigurationFromResource(
+                "lock/one-specified-timeout.yml")))
+                .isEqualTo(BlockingTimeouts.scaleForErrorMargin(86400 * MILLIS_PER_SECOND));
```
nit: SECONDS_IN_ONE_DAY

jeremyk-91 replied:
Right, changed.

hsaraogi commented on timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/lock/BlockingTimeoutsIntegrationTest.java:
```diff
@@ -0,0 +1,71 @@
+    @Test
+    public void returnsDefaultBlockingTimeoutIfConnectorWithoutSpecifiedIdleTimeoutIsMinimal() throws IOException {
+        assertThat(BlockingTimeouts.getBlockingTimeout(OBJECT_MAPPER, getConfigurationFromResource(
+                "lock/one-specified-one-implicit-timeout.yml")))
```
rename the config file to one-default-one-higher-explicit-timeout or something, basically saying 90s > default.

jeremyk-91 replied:
Ah, alright. Yep, agree that we should emphasise that the non-default timeout is higher!

hsaraogi commented on timelock-server/src/test/java/com/palantir/atlasdb/timelock/lock/BlockingTimeLimitedLockServiceTest.java:
```diff
@@ -0,0 +1,131 @@
+
+public class BlockingTimeLimitedLockServiceTest {
+    private static final long BLOCKING_TIME_LIMIT_MILLIS = 10L;
+    private static final long TIME = 42L;
```
`TEST_CURRENT_TIME_MILLIS`

jeremyk-91 replied:
Sure, should be consistent with the above (and that's more descriptive with what it's used for as well)

hsaraogi commented on timelock-server/src/test/java/com/palantir/atlasdb/timelock/lock/BlockingTimeoutsTest.java:
```diff
@@ -0,0 +1,67 @@
+    public void scaleForErrorMarginReducesTimeoutSlightly() {
+        assertThat(BlockingTimeouts.scaleForErrorMargin(ONE_MILLION))
+                .isLessThan(ONE_MILLION)
+                .isEqualTo(970_000);
```
Calculate this from the ERROR_MARGIN constant, after making it `visibleForTesting`?

jeremyk-91 replied:
Sure. It achieves the desired behaviour (if the scale method changes then the test should fail) and would be clearer as well.

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeLimitedLockService.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Override
+    public boolean unlock(LockRefreshToken token) {
```
Hmm, it might be fairly high lift to do this actually, since the methods in question aren't interruptible (so even if you pass them through the time-limiter they will keep going)! We haven't seen this occur in production so I doubt it would be a problem.

My main concern is perf related (though might be a premature optimisation) - because firing off such a request actually spins up a separate thread to run the task (and interrupting the thread if it goes overtime). Happy to discuss this one offline.

hsaraogi replied:
Yes, did realise that this will be more involved to address. Also, we definitely have other workarounds like increasing the idleTimeout if we see the issue in field. I think it is only worth addressing if appears in some production stack going forward.

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeouts.java:
```diff
@@ -0,0 +1,96 @@
+
+    public static final long DEFAULT_IDLE_TIMEOUT = 30 * 1000;
+
+    private static final double ERROR_MARGIN = 0.03;
```
I... did not. This should be configurable, yes :)

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeouts.java:
```diff
@@ -0,0 +1,96 @@
+                .map(connectorDatum -> connectorDatum.get(IDLE_TIMEOUT))
+                .map(idleTimeout -> Duration.parse(idleTimeout).toMilliseconds())
+                .min(Long::compareTo)
+                .orElse(DEFAULT_IDLE_TIMEOUT);
```
Hmm, yep. This would mean there were no connectors in the serialized JSON which seems impossible (I _think_ - since DW will have initialised these when starting the server - but probably worth checking). Will separate the "unknown" logging to its own method and have that do the work.

hsaraogi commented on timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/PaxosTimeLockServerIntegrationTest.java:
```diff
@@ -273,14 +273,14 @@ public void fastForwardToThePastDoesNothing() {
     public void returnsNotFoundOnQueryingNonexistentClient() {
         RemoteLockService lockService = getLockService(NONEXISTENT_CLIENT);
```
`nonExistentLockService` for consistency.

jeremyk-91 replied:
Sure, renamed.

hsaraogi commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/config/TimeLimiterConfiguration.java:
```diff
@@ -0,0 +1,71 @@
+    protected void check() {
+        if (enableTimeLimiting()) {
+            double errorMargin = blockingTimeoutErrorMargin();
+            Preconditions.checkState(
```
nit: maybe combine to a single state check, must be between 0 and 1.

jeremyk-91 replied:
Alright, combined to 1 check.

hsaraogi commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/config/TimeLimiterConfiguration.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    public static TimeLimiterConfiguration getDefaultConfiguration() {
+        return ImmutableTimeLimiterConfiguration.builder()
+                .enableTimeLimiting(false)
```
I would strongly argue that this should be `true`.

jeremyk-91 commented on docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst:
```diff
@@ -131,6 +131,44 @@ default values.
+significantly, if these requests are blocked for longer than the idle timeout of the server's application connector
+on HTTP/2, then Jetty will send a stream closed message to the client. This can lead to an infinite buildup of threads
+and was the root cause of issue `#1680 <https://github.com/palantir/atlasdb/issues/1680>`__. We thus reap the thread
+before the timeout expires, and send a ``BlockingTimeoutException`` to the client indicating that it should retry.
```
Yeah, good point. I've explicitly stated interruptible only, and explained the semantics of the exception (i.e. your request has timed out, but you are free to retry on me).

jeremyk-91 commented on docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst:
```diff
@@ -131,6 +131,44 @@ default values.
+before the timeout expires, and send a ``BlockingTimeoutException`` to the client indicating that it should retry.
+
+This mechanism can be switched on and off, and the time interval between generating the ``BlockingTimeoutException``
+and the actual idle timeout becoming active is configurable. Note that even if we lose the race between generating
```
Yeah, agreed. Changed

jeremyk-91 commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -63,6 +63,19 @@ develop
+         - The format of serialised exceptions occurring on a remote host has been brought in line with that of the `palantir/http-remoting <https://github.com/palantir/http-remoting>`__ library.
+           This should generally improve readability and also allows for more meaningful messages to be sent; we would previously return message bodies with no content for some exceptions (such as ``NotCurrentLeaderException``).
+           (`Pull Request 1 <https://github.com/palantir/atlasdb/pull/1831>`__,
+           `Pull Request 2 <https://github.com/palantir/atlasdb/pull/1783>`__)
```
Good spot! Fixed.

jeremyk-91 commented on docs/source/release_notes/release-notes.rst:
```diff
@@ -63,6 +63,19 @@ develop
+           Previously, these requests would be failed with a HTTP-level exception that the stream was closed.
+           We have rewritten clients constructed via ``AtlasDbHttpClients`` to account for this new behaviour, but custom clients directly accessing the lock service may be affected.
+           This feature is disabled by default, but can be enabled following the TimeLock server configuration :ref:`docs <timelock-server-time-limiting>`.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1783>`__)
```
Yep, changed it. Thanks!

jeremyk-91 commented:
Addressed all comments apart from the default true/false one.

I agree that in terms of production stacks we should definitely set it to true - so the issue is whether this should be set in the timelock server itself, or in our internal product management system. I'm open to discussion either way - let's speak offline (maybe with the rest of the team).

hsaraogi commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeouts.java:
```diff
@@ -0,0 +1,105 @@
+    public static final long DEFAULT_IDLE_TIMEOUT = 30 * 1000;
+
+    private static final String SERVER = "server";
+    private static final String APPLICATION_CONNECTORS = "applicationConnectors";
```
Anyway to deserialize using https://github.com/dropwizard/dropwizard/blob/4705f38cccddf9f15cb95c6f1db8c15a600c8078/dropwizard-core/src/main/java/io/dropwizard/server/DefaultServerFactory.java.

jeremyk-91 replied:
I'll look into this.

jeremyk-91 replied:
Hmm, it seems that the idle timeout isn't serialized and the connector factory doesn't give an easy way of retrieving it without creating a dummy connection, so I think it's easier to stick with this. Thanks for the idea though!

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeLimitedLockService.java:
```diff
@@ -0,0 +1,224 @@
+            return timeLimiter.callWithTimeout(callable, blockingTimeLimitMillis, TimeUnit.MILLISECONDS, true);
+        } catch (InterruptedException e) {
+            // In this case, the thread was interrupted for some other reason, perhaps because we lost leadership.
+            // This is fine but it's probably worth logging.
```
Sure, the line after that should tell people this 🙂

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeouts.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public static long getBlockingTimeout(ObjectMapper mapper, TimeLockServerConfiguration configuration) {
+        // TODO (jkong): Need to decide if this is worth the pain to enforce our invariant, or if we should just
```
Hmm, the alternative I thought would be to have users specify a duration, and we take no responsibility for how they set the value, actually. Not likely for Dropwizard to change these without Jetty changing them!

hsaraogi replied:
But then a user can set it to a value higher than the idleTimeout.
So, this is what can happen:
1. Client tries `/lock/lock/..`
2. jetty times-out and resets the stream
3. The timelimiter is still running the lock operation and succeeds
4. The client retries the lock request and is again waiting on lock expiration.
This can result in kind of a deadlock.

jeremyk-91 replied:
Yep, that's right. That's what we faced in #1680 (essentially not having a timeout = timeout of infinity) - which is why I said _we take no responsibility_ (probably should have said _it is the user's responsibility to set this correctly_). I think the current solution is worth the complexity.

hsaraogi commented on timelock-server/src/test/java/com/palantir/atlasdb/timelock/lock/BlockingTimeoutsTest.java:
```diff
@@ -0,0 +1,87 @@
+    @Test
+    public void scaleForErrorMarginDoesNotReduceAlreadySmallTimeouts() {
+        assertThat(BlockingTimeouts.scaleForErrorMargin(FIVE, TEST_ERROR_MARGIN))
+                .isGreaterThan(0)
```
do we really need this?

jeremyk-91 replied:
You're right - in this case we don't since we're testing that it's equal to a constant already

hsaraogi commented:
This will need an ete test maybe. Could you file an issue for that, maybe just backlog it if we have a smoke test that gives us full confidence?

jeremyk-91 commented:
Yep. I'll include details of the smoke test on the ticket as well.

jeremyk-91 commented on timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/BlockingTimeLimitedLockService.java:
```diff
@@ -0,0 +1,223 @@
+                    specification.method(),
+                    specification.client(),
+                    specification.lockRequest());
+            log.warn(errorMessage, e);
```
We don't expect this to happen too often for most products, though for the shopping product it happened easily 6-12 times per day. Should this be `INFO` instead, to avoid alarming users?

tboam replied:
It doesn't indicate a problem, we expect them to time out occasionally.  Seeing a small number of these in the logs isn't an issue but maybe something we'd investigate to find ways to optimise products.  This points to INFO.

If this is happening a huge number of times then it is a problem, but I expect this becomes clear from the volume.

If we downgrade to info you should also change the variable to message rather than errorMessage, otherwise you're just moving the alarm to when we read the code!

jeremyk-91 replied:
Downgraded to INFO level, and changed to message. Yep!

If it happens a lot, as you've said we will see a large number of logs and that should be sufficient.

jeremyk-91 commented on docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst:
```diff
@@ -131,6 +131,46 @@ default values.
+on HTTP/2, then Jetty will send a stream closed message to the client. This can lead to an infinite buildup of threads
+and was the root cause of issue `#1680 <https://github.com/palantir/atlasdb/issues/1680>`__. We thus reap the thread
+for interruptible requests before the timeout expires, and send an exception to the client indicating that its request
+has timed out, but it is free to retry on the same node. Note that this issue may still occur if a *non-interruptible*
```
Yep, I _think_ the only interruption point should be when blocked on the lock-sync.

jeremyk-91 commented on atlasdb-config/src/main/java/com/palantir/atlasdb/http/ExceptionMappers.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private static SerializableError createSerializableError(Exception exception) {
+        return SerializableError.of(
```
It will. We seemed to run into some issues with serializing the stacktraces which was why I initially expunged it, but figured out it had to deal with the Jdk 8 module - added that.

jeremyk-91 commented on timelock-server/src/integTest/resources/lock/one-default-one-higher-explicit-timeout.yml:
```diff
@@ -0,0 +1,23 @@
+      port: 8080
+    - type: https
+      port: 808
+      idleTimeout: 90 seconds
```
Yes... I wouldn't mind bumping it in our internal recommended configurations. (The normal defaults are a Dropwizard/Jetty thing)

jeremyk-91 commented on docs/source/configuration/external_timelock_service_configs/timelock_server_config.rst:
```diff
@@ -131,6 +131,46 @@ default values.
+
+This mechanism can be switched on and off, and the time interval between generating the ``BlockingTimeoutException``
+and the actual idle timeout is configurable. Note that even if we lose the race between generating this exception and
+the idle timeout, we will retry on the same node. Even if this happens 3 times in a row we are fine, since we will fail
```
(We spoke about the race condition safety offline. If Jetty wins 3x we'll fail over to the other nodes, which will promptly tell us we're not the leader - so we'll go back to the start!)

Lock fairness - yes, I'll add a note here.

jeremyk-91 commented:
I'm now okay with merging this in - @jboreiko @tboam any final comments before we go ahead?

I don't think there's a rush to get this in for 0.40.

jeremyk-91 commented:
(Discussed with John offline, and passed internal smoke tests - going in after CI!)
